### DISCUSSION TITLE
Large fonts - Writer library with Waveshare 1.5 inch B&W OLED display (128x128)

### DISCUSSION THREAD

**Discussion by Treadbrook** [micropython/micropython#9815] [Libraries & Drivers]
Has anyone found a driver for the Waveshare 128x128 B&W OLED that will work with Writer and allow larger fonts?

Or, if not using Writer, does anyone have another way of getting large fonts working with the SSD1327-based OLED displays?

Thanks!

**Comment by peterhinch** [comment]
So far as I can tell Waveshare don't support MicroPython for this display.

There is [this driver](https://github.com/mcauser/micropython-ssd1327/blob/master/ssd1327.py) from Mike Causer, however it would need a little adaptation to work with `Writer` / `CWriter`. The main task is to adapt it so that SSD1327 is subclassed from `framebuf.FrameBuffer` (i.e. use inheritance rather than composition).

As a very general hint for anyone else reading this, the easy way to do this kind of thing is first to define your requirements. Then to select the high level graphics code you want to use (Writer, nano-gui, LVGL or whatever). Then to choose a display which meets requirements and is supported by the software. This avoids getting into realms of writing or adapting display drivers.

**Reply by Treadbrook** [reply]
Thanks for the response, Peter.  Yeah, I started trying to modify Mike Causer's driver using the 1306 driver as a bit of a guideline, but it quickly highlighted my lack of knowledge and so I've put it aside for now.  

I've ordered one of these: https://www.waveshare.com/product/displays/oled/oled-3/1.5inch-rgb-oled-module.htm which is 1351-based and will hopefully work with Writer and your font converter.  If not I will have another crack at the 1327 when I am more awake and alert!

Cheers, Philip

**Reply by mcauser** [reply]
Added FB inheritance version:
https://github.com/mcauser/micropython-ssd1327/blob/writer/ssd1327_fb.py
Untested as I'm moving house and all of my gear is in storage.
Let me know if you have any luck and I'll merge it in.

**Reply by peterhinch** [reply]
That was quick, Mike! 

@Treadbrook there is a little more to do because this is logically a color display because it handles multiple grey levels. Please see [this ref](https://github.com/peterhinch/micropython-nano-gui/blob/master/DRIVERS.md#7-writing-device-drivers). The file `boolpalette.py` is needed to enable 1-bit font files to be blitted onto N-bit color buffers.

**Reply by Treadbrook** [reply]
That is fantastic thanks Michael and Peter!  

Using Mike's new driver and inserting the additions Peter supplied I now have large fonts on the 128x128 Waveshare display and it looks really good.  The only thing I have left to sort out is the brightness... it seems to load the image at about 50% on the grey-scale, so I will have a poke around and see if I can see why.

I *really* appreciate your efforts, guys, I wish I could buy you both a beer. :-)

Cheers, Philip

**Reply by Treadbrook** [reply]
Ahh - scrub that grey-scale comment - looks like a power supply issue.  Rows with a lot of pixels active are dimmer.  This works beautifully.

**Reply by peterhinch** [reply]
If you have a working driver can I suggest you raise a PR against [nano-gui](https://github.com/peterhinch/micropython-nano-gui) to include it. If you're not confident of doing this, please post the code and I will include it.

**Reply by Treadbrook** [reply]
I don't actually know my way around Github at all, sorry Peter, I only set up my account a couple of days ago.  I'm happy to send you the code and you can make it available to everyone.

**Reply by peterhinch** [reply]
Yes please. I will of course credit you and Mike Causer in the code comments.

**Reply by Treadbrook** [reply]
Ummm - so how to send it?  Just paste it here, or is there some other way?  (Sorry, been so busy on the project itself that I really haven't looked at Github any further, I will do soon!)

**Reply by peterhinch** [reply]
Paste it here is simplest. Surround the code with triple backtick characters (`) to preserve indentation.

**Reply by Treadbrook** [reply]
```
"""
MicroPython SSD1327 OLED I2C driver
https://github.com/mcauser/micropython-ssd1327

MIT License
Copyright (c) 2022 Mike Causer

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

__version__ = '1.1.2'

from micropython import const
from framebuf import FrameBuffer, GS4_HMSB
from boolpalette import BoolPalette

# commands
SET_COL_ADDR          = const(0x15)
SET_SCROLL_DEACTIVATE = const(0x2E)
SET_ROW_ADDR          = const(0x75)
SET_CONTRAST          = const(0x81)
SET_SEG_REMAP         = const(0xA0)
SET_DISP_START_LINE   = const(0xA1)
SET_DISP_OFFSET       = const(0xA2)
SET_DISP_MODE         = const(0xA4) # 0xA4 normal, 0xA5 all on, 0xA6 all off, 0xA7 when inverted
SET_MUX_RATIO         = const(0xA8)
SET_FN_SELECT_A       = const(0xAB)
SET_DISP              = const(0xAE) # 0xAE power off, 0xAF power on
SET_PHASE_LEN         = const(0xB1)
SET_DISP_CLK_DIV      = const(0xB3)
SET_SECOND_PRECHARGE  = const(0xB6)
SET_GRAYSCALE_TABLE   = const(0xB8)
SET_GRAYSCALE_LINEAR  = const(0xB9)
SET_PRECHARGE         = const(0xBC)
SET_VCOM_DESEL        = const(0xBE)
SET_FN_SELECT_B       = const(0xD5)
SET_COMMAND_LOCK      = const(0xFD)

# registers
REG_CMD  = const(0x80)
REG_DATA = const(0x40)

class SSD1327(FrameBuffer):
    lut = bytearray(32)   
    def __init__(self, width=128, height=128):
        self.width = width
        self.height = height
        self.buffer = bytearray(self.width * self.height // 2)
        mode = GS4_HMSB
        self.palette = BoolPalette(mode)       
        super().__init__(self.buffer, 128, 128, GS4_HMSB)

        self.col_addr = ((128 - self.width) // 4, 63 - ((128 - self.width) // 4))
        # 96x96     (8, 55)
        # 128x128   (0, 63)

        self.row_addr = (0, self.height - 1)
        # 96x96     (0, 95)
        # 128x128   (0, 127)

        self.offset = 128 - self.height
        # 96x96     32
        # 128x128   0

        self.poweron()
        self.init_display()

    def init_display(self):
        for cmd in (
            SET_COMMAND_LOCK, 0x12, # Unlock
            SET_DISP, # Display off
            # Resolution and layout
            SET_DISP_START_LINE, 0x00,
            SET_DISP_OFFSET, self.offset, # Set vertical offset by COM from 0~127
            # Set re-map
            # Enable column address re-map
            # Disable nibble re-map
            # Horizontal address increment
            # Enable COM re-map
            # Enable COM split odd even
            SET_SEG_REMAP, 0x51,
            SET_MUX_RATIO, self.height - 1,
            # Timing and driving scheme
            SET_FN_SELECT_A, 0x01, # Enable internal VDD regulator
            SET_PHASE_LEN, 0x51, # Phase 1: 1 DCLK, Phase 2: 5 DCLKs
            SET_DISP_CLK_DIV, 0x01, # Divide ratio: 1, Oscillator Frequency: 0
            SET_PRECHARGE, 0x08, # Set pre-charge voltage level: VCOMH
            SET_VCOM_DESEL, 0x07, # Set VCOMH COM deselect voltage level: 0.86*Vcc
            SET_SECOND_PRECHARGE, 0x01, # Second Pre-charge period: 1 DCLK
            SET_FN_SELECT_B, 0x62, # Enable enternal VSL, Enable second precharge
            # Display
            SET_GRAYSCALE_LINEAR, # Use linear greyscale lookup table
            SET_CONTRAST, 0x7f, # Medium brightness
            SET_DISP_MODE, # Normal, not inverted
            SET_COL_ADDR, self.col_addr[0], self.col_addr[1],
            SET_ROW_ADDR, self.row_addr[0], self.row_addr[1],
            SET_SCROLL_DEACTIVATE,
            SET_DISP | 0x01): # Display on
            self.write_cmd(cmd)
        self.fill(0)
        self.write_data(self.buffer)

    def poweroff(self):
        self.write_cmd(SET_FN_SELECT_A)
        self.write_cmd(0x00) # Disable internal VDD regulator, to save power
        self.write_cmd(SET_DISP)

    def poweron(self):
        self.write_cmd(SET_FN_SELECT_A)
        self.write_cmd(0x01) # Enable internal VDD regulator
        self.write_cmd(SET_DISP | 0x01)

    def contrast(self, contrast):
        self.write_cmd(SET_CONTRAST)
        self.write_cmd(contrast) # 0-255

    def rotate(self, rotate):
        self.poweroff()
        self.write_cmd(SET_DISP_OFFSET)
        self.write_cmd(self.height if rotate else self.offset)
        self.write_cmd(SET_SEG_REMAP)
        self.write_cmd(0x42 if rotate else 0x51)
        self.poweron()

    def invert(self, invert):
        self.write_cmd(SET_DISP_MODE | (invert & 1) << 1 | (invert & 1)) # 0xA4=Normal, 0xA7=Inverted

    def show(self):
        self.write_cmd(SET_COL_ADDR)
        self.write_cmd(self.col_addr[0])
        self.write_cmd(self.col_addr[1])
        self.write_cmd(SET_ROW_ADDR)
        self.write_cmd(self.row_addr[0])
        self.write_cmd(self.row_addr[1])
        self.write_data(self.buffer)

    def write_cmd(self):
        raise NotImplementedError

    def write_data(self):
        raise NotImplementedError


class SSD1327_I2C(SSD1327):
    def __init__(self, width, height, i2c, addr=0x3c):
        self.i2c = i2c
        self.addr = addr
        self.cmd_arr = bytearray([REG_CMD, 0])  # Co=1, D/C#=0
        self.data_list = [bytes((REG_DATA,)), None]
        super().__init__(width, height)

    def write_cmd(self, cmd):
        self.cmd_arr[1] = cmd
        self.i2c.writeto(self.addr, self.cmd_arr)

    def write_data(self, data_buf):
        self.data_list[1] = data_buf
        self.i2c.writevto(self.addr, self.data_list)


class SEEED_OLED_96X96(SSD1327_I2C):
    def __init__(self, i2c):
        super().__init__(96, 96, i2c)

    def lookup(self, table):
        # GS0 has no pre-charge and current drive
        self.write_cmd(SET_GRAYSCALE_TABLE)
        for i in range(0,15):
            self.write_cmd(table[i])

class WS_OLED_128X128(SSD1327_I2C):
    def __init__(self, i2c, addr=0x3c):
        super().__init__(128, 128, i2c, addr)
```

**Reply by peterhinch** [reply]
Thank you. There is a small issue.

This is our first greyscale chip and my docs were unclear on the subject of 4-bit drivers. Existing 4-bit drivers map a set of 16 colors onto 4-bit indices into a LUT which is expanded, typically to RGB565, at runtime. This conserves RAM on large displays. This chip supports 4-bit colors natively so there is no need for a LUT.

However there is a need for a static RGB method which maps 24 bit `(r, g, b)` values onto 4-bit greyscale. The [adapted driver](https://github.com/peterhinch/micropython-nano-gui/blob/master/drivers/ssd1327/ssd1327.py) has this. The sample code [here](https://github.com/peterhinch/micropython-font-to-py/blob/master/writer/WRITER.md#23-example-color-code) could be adapted to run. As written `RED` and `GREEN` will both show full intensity but reducing the luminance should now work, e.g.
```python
RED = SSD.rgb(63,0,0)  # 25% luminance
```
Perhaps you could verify that this works.

The docs are now updated to support this driver. I've also clarified [this section](https://github.com/peterhinch/micropython-nano-gui/blob/master/DRIVERS.md#72-color-and-color-compatible-drivers) to describe greyscale drivers.

All comments on code or docs are welcome.

**Comment by peterhinch** [comment]
There is every chance your SSD1351 device will work

It is worth noting that because a display uses the same driver as a supported unit, and has the same resolution, it is still not guaranteed to work out of the box. That is because there can be different ways for the board manufacturer to wire the controller to the display, necessitating changes to the configuration.

However I don't think I've come across this with the SSD1351. There is a simple test described [here](https://github.com/peterhinch/micropython-nano-gui#23-verifying-hardware-configuration). If it passes this, you can be highly confident that it will work with all my libraries.

**Comment by Treadbrook** [comment]
Okay, thanks Peter, I’ll have a look when I get home this evening.


--REPLY above this line to respond--

Philip Adamson
Managing Director
OutSource IT Limited<http://www.outsource.co.nz>

NOTICE:   This e-mail message is the property of the sender. The contents of this message and any attachments are confidential and may be legally privileged. It is intended solely for the addressee. Access  to this e-mail message by anyone else is unauthorized. If you are not an intended recipient (or responsible for delivery of the message to such person), any disclosure, copying, distribution or any action taken or omitted to be taken in reliance  on it, is prohibited and may be unlawful. In such a case, please forward this e-mail message to the sender and delete the message and its attachments from your computer.

**Comment by Treadbrook** [comment]
Sorry about the delay – yes, that works.

Both drivers are quite slow to update… I guess that’s just the amount of data going across the I2C.

**Comment by peterhinch** [comment]
I2C is very slow. Most displays use SPI and it's commonplace to run this at 30-60MHz depending on chip specs (at both ends).

**Comment by Treadbrook** [comment]
Okay, thanks – I will try it using SPI, the display will do both, I’d just designed the PCB around I2C because I was originally intending to use a different display.  It’s only version 1 of the board, I am sure there’ll be other changes as I flesh it out.

Thanks again for your amazing help!  I have learned a lot in the past couple of weeks.

**Comment by Treadbrook** [comment]
For some reason I thought Mike’s driver had SPI capability as well, but it didn’t, so I have added it and the display is staggeringly fast!

I’ll test it a bit and then upload the code for you.

**Comment by rami6711** [comment]
I made sizable font for FrameBuffer: https://github.com/rami6711/fb_plus
